src/ucp/core/ucp_context.c

```diff
@@ -22,13 +22,53 @@
 #include <ucs/debug/debug_int.h>
 #include <ucs/sys/compiler.h>
 #include <ucs/sys/string.h>
+#include <ucs/type/init_once.h>
 #include <ucs/vfs/base/vfs_cb.h>
 #include <ucs/vfs/base/vfs_obj.h>
 #include <string.h>
 
 
 #define UCP_RSC_CONFIG_ALL    "all"
 
+#define UCP_AM_HANDLER_FOREACH(_macro) \
+    _macro(UCP_AM_ID_SINGLE) \
+    _macro(UCP_AM_ID_FIRST) \
+    _macro(UCP_AM_ID_MIDDLE) \
+    _macro(UCP_AM_ID_SINGLE_REPLY) \
+    _macro(UCP_AM_ID_ATOMIC_REQ) \
+    _macro(UCP_AM_ID_ATOMIC_REP) \
+    _macro(UCP_AM_ID_PUT) \
+    _macro(UCP_AM_ID_GET_REQ) \
+    _macro(UCP_AM_ID_GET_REP) \
+    _macro(UCP_AM_ID_CMPL) \
+    _macro(UCP_AM_ID_STREAM_DATA) \
+    _macro(UCP_AM_ID_EAGER_ONLY) \
+    _macro(UCP_AM_ID_EAGER_FIRST) \
+    _macro(UCP_AM_ID_EAGER_MIDDLE) \
+    _macro(UCP_AM_ID_EAGER_SYNC_ONLY) \
+    _macro(UCP_AM_ID_EAGER_SYNC_FIRST) \
+    _macro(UCP_AM_ID_EAGER_SYNC_ACK) \
+    _macro(UCP_AM_ID_OFFLOAD_SYNC_ACK) \
+    _macro(UCP_AM_ID_WIREUP) \
+    _macro(UCP_AM_ID_RNDV_RTS) \
+    _macro(UCP_AM_ID_RNDV_ATS) \
+    _macro(UCP_AM_ID_RNDV_ATP) \
+    _macro(UCP_AM_ID_RNDV_RTR) \
+    _macro(UCP_AM_ID_RNDV_DATA)
+
+#define UCP_AM_HANDLER_DECL(_id) extern ucp_am_handler_t ucp_am_handler_##_id;
+
+#define UCP_AM_HANDLER_ENTRY(_id) \
+    ucp_am_handlers[_id].features = ucp_am_handler_##_id.features; \
```

Comment:
maybe ucp_am_handlers could be array of pointers, and point directly to &ucp_am_handler_##_id?
even make it const pointer if possible

Reply:
> maybe ucp_am_handlers could be array of pointers, and point directly to &ucp_am_handler_##_id?
> even make it const pointer if possible

yep, done